contracts/rosca.sol

```diff
@@ -1,266 +1,258 @@
-        {
-            members[requestor] = USER({paid: 0 , contributed: 0, alive: true, allowance: 0});
-            memberAddress.push(requestor);
-            numRounds++;
-            LogParticipantRegistered(requestor);
-            delete(pendingJoinRequest[requestor]); // take out the requestor's address in the pending list
-        }
-        else throw;
+      lowestBid = contributionSize;
+      winnerAddress = 0;
 
-
-    }
-
-    // v2: function leave();
-
-    /**
-     * Processes a periodic contribution from msg.sender (the user-proxy).
-     * Any excess funds will be withdrawable through withdraw().
-     */
-    function contribute() payable
+      currentRound++;
+    }else if(currentEpoch < numEpochs) //end of Epoch reset rounds
     {
-        //Product condsideration : are we gonna limit how much ether could be sent? is it total, or how much sent each time?
-        if(members[msg.sender].alive)
-        {
-            members[msg.sender].contributed = msg.value;
-            totalAmountInContract = msg.value;
-            //TODO(shine) : look into security about deposit and withdrawal
-            LogContributionMade(msg.sender, msg.value);
-        }
-        else
-           throw;
+      currentEpoch++;
+      currentRound = 1;
+      lowestBid = contributionSize;
     }
-
-    /**
-     * Registers a bid from msg.sender. If msg.sender has already won a round,
-     * this method will do nothing (or throw?).
-     */
-    function bid(uint distrubtionAmountInWei)
+  }
+  /**
+   * Anyone not already a member of ROSCA can request to join and they'll be put into
+   * pendingJoinRequest untill foreman accept request or ROSCA has started
+   *
+   */
+  function joinRequest() beforeStart
+  {
+    //only put the request in the pending list if they are not in the ROSCA already
+    if(!members[msg.sender].alive)
+      pendingJoinRequest[msg.sender] = true;
+    else
+      throw;
+  }
+  /**
+   * foreman must pass in the address of the requtor to approve the ROSCA participation
+   * once a requestor had been registered as member, the address will be taken out of pendingJoinRequest
+   **/
+  function acceptJoinRequest(address requestor)
+    onlyForeman
+    beforeStart
+  {
+    if(pendingJoinRequest[requestor])
     {
-        if(distrubtionAmountInWei < lowestBid || !(members[msg.sender].paid < currentEpoch))
-        {
-            //set new lowestBid and winnerAddress, also trigger event new Bid placed
-            lowestBid = distrubtionAmountInWei;
-            winnerAddress = msg.sender;
-            LogNewLowestBid(lowestBid, winnerAddress);
-
-        }else
-            throw;
+      members[requestor] = USER({paid: 0 , contributed: 0, alive: true, pendingWithdrawl: 0});
+      membersAddresses.push(requestor);
+      numRoundsInEpoch++;
+      LogParticipantRegistered(requestor);
+      delete(pendingJoinRequest[requestor]); // take out the requestor's address in the pending list
```

Comment:
i understand what you mean by deleting all the remaining request, i dont get the part about gas refund to the caller